### ISSUE TITLE
Revert text setter ergonomics pessification; add raw_set methods

### ISSUE THREAD

**Issue by btj** [capnproto/capnproto-rust#470]
I just upgraded to 0.18 and I'm bitten by #433 . I need to uglify all of my text setter calls with `.into()` now ☹️ . In fact, since most of them set from a String (say `x`), what used to be `set_foo(&x)` now has to become `set_foo(x.as_str().into())`.

How about we revert `set_foo` to taking a `&str` and have a separate `raw_set_foo` that takes a `&[u8]` or a Reader or whatever?

I understand one motivation for having text setters continuing to take Readers was consistency. But what is more important: consistency with standard Rust usage of representing text as `&str` or consistency with the rest of capnp-rust?

**Comment by dwrensha** [comment]
Note that instead of `set_foo(x.as_str().into())`, you can do `set_foo(x[..].into())`, which is a bit shorter.

**Comment by dwrensha** [comment]
Adding separate setter methods for `capnp::Text::Reader` and `str` could maybe work. We would need to figure out:
 1. naming -- I don't like `raw_` because that already means "raw pointer" elsewhere
 2. how it would interact with generics -- if a field has a generic type `T`, then does that field now need two different setters, because `T` might be instantiated to `text::Owned`?

Alternatively, we could consider reviving #421, which would make pointer setters accept any type that implements `SetPointerBuilder`.

I'm reluctant to do either of these because they add more complexity, but maybe the potential ergonomic gains for downstream users outweighs the cost.

**Comment by btj** [comment]
Alternative naming suggestion: for a text field `foo` and text reader `r`, `fd_foo().set(r)`. `fd_foo` returns a `TextField`.

This aligns with another proposal I was planning to make, which is to reduce the boilerplate of building lists of structs, by enabling writing, for a field `bars` of Capnp type `List(Bar)` and some `v` of native Rust type `Vec<MyBar>`, something like `fd_bars().from_iterable(v, |bar_cpn, bar| { /* Initialize the bar_cpn from the bar */ })`. The essence of a Field is that, given a length, it can return a Builder: `fd_bars().init(len)` would return a BarBuilder. Conveniences such as `from_iterable` can be built on top of that.

**Comment by btj** [comment]
FYI: I decided to vendor capnproto-rust, so I am not blocked on this issue being resolved here. I reverted the text setters to taking `&str` https://github.com/verifast/capnproto-rust/commit/01871b5d9a86412f7ff2e192a02c8ab450d2a56b . I also changed the `init_foo` methods for initializing lists to taking a `usize` length instead of `u32`, and the list `get` method to taking a `usize` index. Furthermore, I generate a `fill_foo` method for each field `foo` of type list-of-struct, that looks like this:
```rust
#[inline]
pub fn fill_args<E, F>(&mut self, elems: E, body: F)
    where F: FnMut(crate::vf_mir_capnp::body::basic_block::operand::Builder<'_>, E::Item), E: IntoIterator, E::IntoIter: ExactSizeIterator {
  <::capnp::struct_list::Builder<'_,crate::vf_mir_capnp::body::basic_block::operand::Owned>>::fill(self.builder.reborrow().get_pointer_field(1), elems, body)
}
```
which uses a new `fill` associated function in `struct_list::Builder`:
```rust
impl<'a, T> Builder<'a, T>
where
    T: crate::traits::OwnedStruct + 'a,
{
    pub fn fill<E, F>(builder: PointerBuilder<'a>, elems: E, mut body: F)
        where F: FnMut(T::Builder<'_>, E::Item), E: IntoIterator, E::IntoIter: ExactSizeIterator
    {
        let iter = elems.into_iter();
        let mut list_builder = Self::init_pointer_with_usize_length(builder, iter.len());
        for (idx, elem) in iter.enumerate() {
            let elem_cpn = list_builder.reborrow().get(idx);
            body(elem_cpn, elem);
        }
    }
}
```
Commit: https://github.com/verifast/verifast/commit/60e9ca43d310551f98f4850cf3f6c35b25602401#diff-fe25a34e136e01802f5f365abad8bf12d11c18a922ffd283bed2c3af52eb1e35

I also did something similar for text lists: I generate a method `fill_foo` for each field of type `List(Text)`:
```rust
#[inline]
pub fn fill_arg_names<E>(&mut self, elems: E)
    where E: IntoIterator, E::IntoIter: ExactSizeIterator, E::Item: AsRef<str> {
  <::capnp::text_list::Builder<'_>>::fill(self.builder.reborrow().get_pointer_field(3), elems)
}
```
which uses a new `fill` associated function on `text_list::Builder`:
```rust
impl<'a> Builder<'a> {
    pub fn fill<E>(builder: PointerBuilder<'a>, elems: E)
        where E: IntoIterator, E::IntoIter: ExactSizeIterator, E::Item: AsRef<str>
    {
        let iter = elems.into_iter();
        let mut list_builder = Self::init_pointer_with_usize_length(builder, iter.len());
        for (idx, elem) in iter.enumerate() {
            list_builder.set(idx, elem.as_ref());
        }
    }
}
```
Commit: https://github.com/verifast/verifast/commit/68b70c64b9c6498b3345e349bfb82922238188a7

**Comment by dwrensha** [comment]
I revived #421 as #472. I think this works well, but unfortunately would re-break the code of everyone who has already updated to capnp-v0.18.

**Comment by dwrensha** [comment]
Closing, because I've landed #472.
Thanks for pushing on this!
